src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -102,19 +102,21 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	CacheService cacheService = new CacheService(dto);
     	MessageCacheItem cacheItem = cacheService.getCacheOfComponent();
     	if (cacheItem != null) { // Item is in cache
-    		if (cacheItem.getCachedData().isEmpty()) { // This means that the data to be used is from a fallback locale.
-				// If expired, try to first create and store cacheItem for the requested locale in a separate thread.
-				if (cacheItem.isExpired())
-					this.createCacheItemTask(null); // Pass null so that locale fallback will not be applied.
-
-    			// Use cached data of cacheItem.getLocale() --> fallback locale
-				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), cacheItem.getLocale(), dto.getProductID(), dto.getVersion());
-				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(null);
-			} else if (cacheItem.isExpired()) {
+    		if (cacheItem.isExpired()) {
     			// Refresh the cacheItem in a separate thread
     			refreshCacheItemTask(cacheItem);
     		}
-    	} else { // Item is not in cache. Create and store cacheItem for the requested locale
+    	} else { // Item is not in cache.
+			ProductService ps = new ProductService(dto);
+			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
+				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
+					return new ComponentService(matchedLocaleDTO).getMessages(null);
```

Comment:
Seems recursive call is not good.

Reply:
There is no recursive call in this line. Notice that there getMessages is invoked on a seaparate instance of ComponentService.

Reply:
Also, this logic (lines 110-117) is not new. I just moved it out of the createCacheItem method that is called in line 120 into here for code clean up.

Reply:
I meant line 115. 'not good' means that it gets the messages twice: one for supported locale, another or not supported locale. The logic is duplicated, the locale should be handled before call getMessages.

Secondly, when i read 'return new ComponentService(matchedLocaleDTO).getMessages(null);', i don't get what it means by passing a null value.

Reply:
This code has been there before this PR, so I suggest we open another issue for your concern. becasue this PR is for code clean up only, not changing any logic.

The intention of this line is exactly what you mentioned. We want to get the matched locale **instead of** the actual requested locale if a match is found. This has been the same strategy for calling the fallback locale.. new ComponentService(fallbacklocaleDTO).getMessages() if needed. What is your suggestion on how to do that?

To understand the behavior when passing null to getMessages, read the javadoc of getMessages:
`/**
	 * Get MessageCacheItem from cache.
	 * The cache is refreshed if MessageCacheItem is expired or not found.
	 *
	 * @param fallbackLocalesIter The locale fallback queue to be used on failure. If null, there will be no fallback mechanism on failure so the message map will be empty.
	 *
	 * @return A MessageCacheItem whose data map is one of the following:
	 * <ul>
	 * 		<li>The messages in the requested locale</li>
	 * 	 	<li>The messages in a fallback locale</li>
	 * </ul>
	 */`